internal/pkg/EAR/ear_state.go

```diff
@@ -0,0 +1,390 @@
+
+// AbsPartitionInfo maintains information at the abstract partition level (which
+// consists of a set of abstract references). It maintains the size of the
+// partition (used for weighted join in union-find) and a AbsReferenceFieldMap,
+// which is a mapping from values F (instance fields, array contents etc) to
+// abstract references R.
+type AbsPartitionInfo struct {
+	numMembers uint                 // number of members
+	fieldRefs  AbsReferenceFieldMap // field mapping
+}
+
+// PartitionInfoMap maps an abstract reference to its partition.
+type PartitionInfoMap map[AbsReference]AbsPartitionInfo
+
+// AbsState is Equivalent Abstract References (EAR) state.
+type AbsState struct {
+	// Map from an abstract ref (a partition representative) to information about
+	// this partition.
+	partitions PartitionInfoMap
+	// Map from an abstract ref to its parent abstract ref.
+	parents ParentMap
+
+	// Set to true after all mutation operations have completed: only lookups will
+	// be performed from here on.
+	finalized bool
+
+	// Mapping from a partition representative to its members. This map is
+	// initialize in Finalize method, which guarantees to avoid duplication.
+	// Users are accessing this information via GetPartitionMembers method,
+	// which does not allow any modification.
+	members map[AbsReference][]AbsReference
+
+	// Map from each field reference to its parent references. For example, record
+	// r->o for "o.x = r". This map is for accelerating external queries on
+	// fields, and is not required by the EAR analysis itself.
+	// It is constructed separately using the "ConstructFieldParentMap()"
+	// at the final phase.
+	revFields map[AbsReference][]AbsReference
+}
+
+// NewAbsState creates an empty abstract state.
+func NewAbsState() *AbsState {
+	return &AbsState{
+		partitions: make(PartitionInfoMap),
+		parents:    make(ParentMap),
+		finalized:  false,
+		members:    make(map[AbsReference][]AbsReference),
+		revFields:  make(map[AbsReference][]AbsReference),
+	}
+}
+
+// HasAbsRef returns true if "aref" is an abstract reference present in the state.
+func (state *AbsState) HasAbsRef(aref AbsReference) bool {
+	_, ok := state.parents[aref]
+	return ok
+}
+
+// GetAllAbsRefs get all abstract references via out-parameter
+// "arefs" (should be non-nil).
+func (state *AbsState) GetAllAbsRefs(arefs AbsReferenceSet) {
+	for k := range state.parents {
+		arefs[k] = true
+	}
+}
+
+// GetNumAbsRefs returns the total number of all abstract references.
+func (state *AbsState) GetNumAbsRefs() int {
+	return len(state.parents)
+}
+
+// GetAllPartitionReps gets all abstract references that are partition
+// representatives via out-parameter "reps" (should be non-nil).
+func (state *AbsState) GetAllPartitionReps(reps AbsReferenceSet) {
+	for k, v := range state.parents {
+		if k == v {
+			reps[k] = true
+		}
+	}
+}
+
+// GetPartitionRep gets the partition representative of abstract reference "aref"
+// ("aref" must belong to this state).
```

Comment:
This is a fairly harsh contract. Is there no reasonable return value for not present?